Add Tests for color.is-in-gamut()

SondraE opened the pull request:
![](https://images.unsplash.com/photo-1472491235688-bdc81a63246e?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwxfDB8MXxyYW5kb218MHx8Y3V0ZSxhbmltYWwsQ0hBTkdFX01FfHx8fHx8MTY3Mjg2MTI0Nw&ixlib=rb-4.0.3&q=80&)


**Description**
(https://github.com/sass/sass-spec/pull/1842)

**Steps to test/reproduce**
- new file called in_gamut
- tests to determine which colors are in gamut

nex3 committed:
[Color 4] Update tests for Color Level 4 (#1846)

This updates most tests for the currently-implemented chunk of the
Color 4 feature. A few tests that need more thorough reworking (such
as being moved out of `error/` for things that are no longer intended
to be errors) are marked as "todo" instead.

See #1828
See sass/sass#2831

nex3 committed:
Use feature branches for testing the JS API (#1853)

This also skips a few color tests for the embedded host

SondraE committed:
Add tests for color.is-legacy() (#1842)

nex3 committed:
Don't auto-skip dart-sass-embedded along with dart-sass (#1857)

I added this under the erroneous impression that we ran *language*
tests against Dart Sass rather than just *JS API* tests.

SondraE committed:
Add tests for color.space() (#1851)

Co-authored-by: dvdherron <dvdherron@gmail.com>
Co-authored-by: Natalie Weizenbaum <nweiz@google.com>

dvdherron committed:
[Color 4] Update tests for color.grayscale() (#1850)

Co-authored-by: Natalie Weizenbaum <nweiz@google.com>

SondraE committed:
Add tests for all Deprecated Functions (#1852)

Co-authored-by: Stacy Kvernmo <stacy@oddbird.net>
Co-authored-by: dvdherron <dvdherron@gmail.com>
Co-authored-by: Natalie Weizenbaum <nweiz@google.com>

SondraE committed:
add tests for color is in gamut

dvdherron commented on spec/core_functions/color/in_gamut:
```diff
@@ -0,0 +1,358 @@
+================================================================================
+<===> display-p3/input.scss
+@use "sass:color";
+a {b: color.is-in-gamut(color(display-p3 -1 0.35 0.3 / 1))}
```
I'm not clear on what the division (`/`) is for in some of the color functions here? I think the `out-of-gamut` colors would still be formatted like other colors. We would just have find actually `out-of-gamut` colors to test.

dvdherron reviewed (commented):
I think this is a good start @SondraE . There's a part missing though. This function also takes a `$space` argument. I don't think these test that part of the function yet. We can talk through this after standup tomorrow if you're around. Thanks for starting this!

SondraE committed:
removed division

dvdherron committed:
fix file type

SondraE committed:
added error for space and tests for space

dvdherron commented on spec/core_functions/color/in_gamut.hrx:
```diff
@@ -0,0 +1,399 @@
+
+<===>
+================================================================================
+<===> rgb/input.scss
```
This is the second test named "rgb/input.scss". There can only be one. Are you testing different things in these two tests? If they are just two different rgb colors you can remove one of the tests.

dvdherron commented on spec/core_functions/color/in_gamut.hrx:
```diff
@@ -0,0 +1,399 @@
+
+<===>
+================================================================================
+<===> rgba/input.scss
```
same here. two rgba tests. 

If you run the command `npm run lint-spec` in your terminal it will show you these errors and sometimes show you how to correct them.

dvdherron commented on spec/core_functions/color/in_gamut.hrx:
```diff
@@ -0,0 +1,399 @@
+================================================================================
+<===> named/input.scss
+@use "sass:color";
+a {b: color.is-in-gamut($color: rgb(255 0 0))}
```
In the `named` test, in addition to `$color` you would also have a parameter for `$space`. That's the second argument in this function. 

I think you could remove this one and used the one currently named `space_named`, but change `space_named` to `named`.

dvdherron commented on spec/core_functions/color/in_gamut.hrx:
```diff
@@ -0,0 +1,399 @@
+
+<===>
+================================================================================
+<===> space_named/input.scss
```
```suggestion
<===> named/input.scss
```

dvdherron commented on spec/core_functions/color/in_gamut.hrx:
```diff
@@ -0,0 +1,399 @@
+@use "sass:color";
+a {b: color.is-in-gamut($color: display-p3(1 1 0), $space: srgb)}
+
+<===> named_space/output.css
```
```suggestion
<===> named/output.css
```

dvdherron reviewed (commented):
The tests

dvdherron committed:
rename to is_in_gamut

SondraE committed:
more error and space conversion tests

dvdherron committed:
cleanup plus update expectations to match dart sass

dvdherron commented on spec/core_functions/color/is_in_gamut.hrx:
```diff
@@ -18,7 +18,7 @@ a {
 <===> rgba/input.scss
 @use "sass:color";
-a {b: color.is-in-gamut(rgba(-1 122 180))}
+a {b: color.is-in-gamut(rgba(-1 122 180 / 0.4))}
```
the rgba function needed an alpha channel.

dvdherron commented on spec/core_functions/color/is_in_gamut.hrx:
```diff
@@ -33,7 +33,7 @@ a {b: color.is-in-gamut(hwb(0 300% 0%))}
 <===> hwb/output.css
 a {
-  b: false;
+  b: true;
```
Reading up on this, but it seems hwb and hsl values are clamped sometimes when they are out of range: https://github.com/sass/sass/issues/3409#issuecomment-1281351223

dvdherron commented on spec/core_functions/color/is_in_gamut.hrx:
```diff
@@ -44,14 +44,14 @@ a {b: color.is-in-gamut(hsl(0 -1% 0%))}
 <===> srgb/input.scss
 @use "sass:color";
-a {b: color.is-in-gamut(color(srgb 2 0.07843 0.823530))}
+a {b: color.is-in-gamut(color(srgb 1.5 0 0))}
```
Just tried to simplify some of the numbers here and it some other places below.

dvdherron commented on spec/core_functions/color/is_in_gamut.hrx:
```diff
@@ -216,7 +216,7 @@ a {
 <===> named/input.scss
 @use "sass:color";
-a {b: color.is-in-gamut($color: display-p3(1 1 0), $space: srgb)}
+a {b: color.is-in-gamut($color: color(display-p3 1 1 0), $space: srgb)}
```
there were extra parentheses

dvdherron commented on spec/core_functions/color/is_in_gamut.hrx:
```diff
@@ -273,15 +273,37 @@ a {
 <===> error/too_many_args/input.scss
 @use "sass:color";
-a {b: color.is-in-gamut(rgb(0 255 0), hwb, hsl)}
+a {b: color.is-in-gamut(rgb(0 255 0), hwb, c)}
```
Placeholder values should just be letters.

dvdherron committed:
change to quoted string error

dvdherron committed:
cleanup and add more errors

dvdherron reviewed (commented):
Looks ready to send upstream for a Sass review!

SondraE committed:
varied which channel in the color functions was out of gamut

dependabot[bot] committed:
Bump json5 from 2.2.0 to 2.2.3 (#1865)

Bumps [json5](https://github.com/json5/json5) from 2.2.0 to 2.2.3.
- [Release notes](https://github.com/json5/json5/releases)
- [Changelog](https://github.com/json5/json5/blob/main/CHANGELOG.md)
- [Commits](https://github.com/json5/json5/compare/v2.2.0...v2.2.3)

---
updated-dependencies:
- dependency-name: json5
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nex3 committed:
Merge remote-tracking branch 'origin/main' into feature.color-4

nex3 committed:
Expand error tests for trailing dots in numbers (#1869)

See https://github.com/sass/dart-sass#1856

nex3 committed:
Fix a race condition in the Dart compiler (#1871)

If Dart chose to block while writing stdout or stderr, only requesting
one of them could cause the runner to deadlock.

nex3 committed:
Merge branch 'main' of github.com:sass/sass-spec into main-into-color-4

nex3 committed:
Merge branch 'feature.color-4' of github.com:sass/sass-spec into main-into-color-4

nex3 committed:
Fix a hue test with the wrong unit

nex3 committed:
Merge remote-tracking branch 'origin/main' into feature.color-4

nex3 committed:
Merge branch 'feature.color-4' of github.com:sass/sass-spec into color-in-gamut
